File: src/main/java/fi/evident/dojolisp/eval/Environment.java
package fi.evident.dojolisp.eval;

import static fi.evident.dojolisp.utils.Objects.requireNonNull;

public final class Environment {
    
    private final Object[] bindings;
    private final Environment parent;

    public Environment(int size) {
        this.bindings = new Object[size];
        this.parent = null;
    }

    private Environment(Object[] bindings, Environment parent) {
        this.bindings = requireNonNull(bindings);
        this.parent = parent;
    }

    public Object lookup(VariableReference var) {
        return bindingsForFrame(var.frame)[var.offset];
    }

    public void set(VariableReference var, Object value) {
        bindingsForFrame(var.frame)[var.offset] = value;
    }

    private Object[] bindingsForFrame(int depth) {
        Environment env = this;
        for (int i = 0; i < depth; i++)
            env = env.parent;
        return env.bindings;
    }

    public Environment extend(Object[] args) {
        return new Environment(args, this);
    }
}

File: src/main/kotlin/komu/blunt/eval/Environment.kt
package komu.blunt.eval

import komu.blunt.analyzer.VariableReference
import java.lang.Math.max
import java.util.*

abstract class Environment {

    operator fun get(v: VariableReference) =
        get(v.frame, v.offset)

    operator fun set(v: VariableReference, value: Any?) {
        set(v.frame, v.offset, value)
    }

    abstract operator fun get(frame: Int, offset: Int): Any?
    abstract operator fun set(frame: Int, offset: Int, value: Any?)

    fun extend(envSize: Int, arg: Any?): Environment {
        val args = Array<Any?>(envSize) { null }
        args[0] = arg
        return NestedEnvironment(args, this)
    }

    fun extend(envSize: Int): Environment =
        NestedEnvironment(Array(envSize) { null }, this)
}

class RootEnvironment : Environment() {

    private var bindings = Array<Any?>(512) { null }

    override fun set(frame: Int, offset: Int, value: Any?) {
        checkFrame(frame)

        bindings[offset] = value
    }

    override fun get(frame: Int, offset: Int): Any? {
        checkFrame(frame)

        return bindings[offset]
    }

    fun define(v: VariableReference, value: Any?) {
        checkFrame(v.frame)

        if (v.offset >= bindings.size)
            bindings = Arrays.copyOf(bindings, max(v.offset + 1, bindings.size * 2))

        bindings[v.offset] = value
    }

    private fun checkFrame(frame: Int) {
        if (frame != 0 && frame != VariableReference.GLOBAL_FRAME)
            throw IllegalArgumentException("invalid frame $frame");
    }
}

class NestedEnvironment(private val bindings: Array<Any?>, private val parent: Environment) : Environment() {

    override fun get(frame: Int, offset: Int): Any? =
        if (frame == 0)
            bindings[offset]
        else
            parent[frame-1, offset]

    override fun set(frame: Int, offset: Int, value: Any?) {
        if (frame == 0)
            bindings[offset] = value
        else
            parent[frame-1, offset] = value
    }
}
